### DISCUSSION TITLE
With keyboard enhancement protocol Alt+Shift bindings no longer work

### DISCUSSION THREAD

**Discussion by mkon** [helix-editor/helix#9208] [Troubleshooting] [answered]
### Summary

Recently my terminal alacritty started supporting the kitty keyboard protocol. I can verify that helix is now using it since there is this message when booting helix with `-vv`: `The keyboard enhancement protocol is supported in this terminal`.

I also have some keybinds like:
```
[keys.normal]
"A-n" = "extend_search_next"
"A-N" = "extend_search_prev"
```

So far this always worked, but now it only runs the lower case binding even when holding shift. Is there a different way to bind keys with the kitty protocol?

In terminal I can run `ctrl-v` and then `alt-shift-n` and it will print `^[N` (without shift it is `^[n`) which looks correct to me. But running `hx` from the same terminal will not recognise the shift.

### Reproduction Steps

 

I tried this:

Configure helix with alt-shift combinations like the example above.

1. `hx`

I expected this to happen:

alt-shift-n to search backwards.

Instead, this happened:

alt-shift-n searched forwards.

### Helix log

Is there a way to get keypress events in logs? `-vv` Was not enough.

### Platform

macOS

### Terminal Emulator

alacritty

### Installation Method

brew

### Helix Version

helix 23.10 (f6021dd0)

**Comment by the-mikedavis** [comment, marked as answer]
I can't reproduce this with Kitty. AFAIK the kitty keyboard protocol support in alacritty is fairly new so this might be an alacritty bug.

The way I typically debug key events is to run our terminal manipulation library's event-read example: https://github.com/crossterm-rs/crossterm/blob/master/examples/event-read.rs (i.e. clone that repository and run `cargo run --example event-read`).

In order to make the kitty keyboard protocol settings the same, you can modify this block: https://github.com/crossterm-rs/crossterm/blob/08762b3ef4519e7f834453bf91e3fe36f4c63fe7/examples/event-read.rs#L78-L86

to be:

```rust
PushKeyboardEnhancementFlags(
    KeyboardEnhancementFlags::DISAMBIGUATE_ESCAPE_CODES
        | KeyboardEnhancementFlags::REPORT_ALTERNATE_KEYS
)
```

Those are the settings we enable in Helix.

**Comment by mkon** [comment]
Hey @the-mikedavis, thanks for the swift response. It seems to be alacritty indeed. It sends these events both both `N` and `n`:

```
Event: Key(KeyEvent { code: Char('n'), modifiers: KeyModifiers(ALT), kind: Press, state: KeyEventState(0x0) })
Event: Key(KeyEvent { code: Char('n'), modifiers: KeyModifiers(ALT), kind: Press, state: KeyEventState(0x0) })
```

While in kitty it is this:

```
Event: Key(KeyEvent { code: Char('n'), modifiers: KeyModifiers(ALT), kind: Press, state: KeyEventState(0x0) })
Event: Key(KeyEvent { code: Char('N'), modifiers: KeyModifiers(ALT), kind: Press, state: KeyEventState(0x0) })
```

I assumed the `ctrl-v` output pointing to helix, but I guess this does not use the advanced keyboard protocol.
